src/cubicbez.rs

```diff
@@ -157,7 +157,15 @@ impl ParamCurveArclen for CubicBez {
                 rec(&c0, accuracy * 0.5, depth + 1) + rec(&c1, accuracy * 0.5, depth + 1)
             }
         }
-        rec(self, accuracy, 0)
+
+        // Check if the bezier curve is degenerate, or almost degenerate
+        // A degenerate curve where all points are identical will cause infinite recursion in the rec function (well, until MAX_DEPTH at least) in all branches.
+        let small = accuracy * 0.00000001;
```

Comment:
Pretty sure this will still lead to incorrect results in some cases, mostly because you're comparing a squared distance with a distance.

Thinking about this, I propose:

`(self.p1 - self.p0).hypot2() + (self.p2 - self.p3).hypot2() <= CONST * accuracy * accuracy`

This has the side effect of short-cutting the case where p0 = p1 and p2 = p3. The only question is, what is CONST? I think 1.0 is adequate, but it would be better to do mathematical analysis.

Reply:
Hmmm.. I think 0.5 will work.
The upper bound for the length of the curve is

```|p1-p0| + |p2 - p1| + |p3 - p2|```

Given fixed values of p0 and p3 (assume wlog that they lie on a straight horizontal line) and your constraint

```|p1 - p0|^2 + |p2 - p3|^2 <= C^2```

Then the largest value of `|p2 - p1|` happens (I postulate) when `p1` is `C/sqrt(2)` units to the left of `p0` and `p2` is `C/sqrt(2)` units to the right of `p2`.

Or in other words `|p2 - p1| <= |p3 - p0| + C*(2*1.0/sqrt(2)) = |p3 - p0| + C*2*sqrt(2)`.

Thus the error

`||p3 - p0| - (|p1-p0| + |p2 - p1| + |p3 - p2|)| <= C*4/sqrt(2)`

Therefore I think we must set `C*4/sqrt(2) = precision` and thus `C^2 = precision*precision*1/8`.

Reply:
Ok, I've been properly nerd-sniped, and wanted to test this empirically. Here's my [code](https://gist.github.com/raphlinus/37a0ed567a9dcc92f16bbecca802ab4a).

From random testing, CONST = 0.8714 seems to do the trick, but it's hard to know if that's a hard bound. Interestingly, the worst case behavior seems to be a "tear drop" shape where p0 and p3 nearly coincide.

```html
<!DOCTYPE html>
<html>
<body>

<svg height="400" width="450">
  <path d="M165.61722290078575 43.216428638303974C5.65289484019762 101.67969576236118 68.23986815921117 182.9261339463449 165.61722290078575 43.216428638303974" stroke="blue" stroke-width="2" fill="none" />
</svg>

</body>
</html>
```

So it's certainly the case that 0.5 is a conservative bound, as you suggest :)

Reply:
Haha :p. That is an interesting worst case, I wouldn't have expected that.
And my 1/8 is very very conservative in that case :p.

I don't think it matters that much which constant is chosen as long as it is good enough. I'd propose 0.5 in that case since it seems to be well under the empirical bound and it's a nice number. However a comment should explain the reasoning.

Reply:
That sounds like a good plan, I am happy with 0.5. Feel free to make the change and I will merge it with thanks.